src/persist-client/src/iter.rs

```diff
@@ -166,9 +165,13 @@ pub(crate) enum ConsolidationPart<T, D> {
         part: EncodedPart<T>,
         cursor: Cursor,
     },
+    /// A part that was not necessarily in sorted order.
+    /// We store the original part along with a bunch of "pointers" into it that _are_ in sorted
+    /// order / in a convenient format for our streaming iterator. (For lifetimey reasons, these
+    /// are stored as [Cursor]s instead of ordinary references.)
     Sorted {
-        data: Vec<((Vec<u8>, Vec<u8>), T, D)>,
-        index: usize,
+        part: EncodedPart<T>,
+        cursors: VecDeque<(Cursor, T, D)>,
```

Comment:
Storing `Cursor`s as the pointers into the `EncodedPart` makes me a little nervous because it feels really easy for someone to call `.pop(...)` instead of `.peek(...)` and accidentally causes the data to become unsorted and invalid. What do you think about wrapping cursor, e.g. something like:

```
/// A newtype wrapper around `Cursor` that can safely be used for pointers into an `EncodedPart`.
pub struct Pointer(Cursor);

impl Pointer {
  pub fn peek(part: &EncodedPart) -> (...) {
    self.0.peek(part)
  }
}
```

Which would avoid the above issue entirely

Reply:
Point!

One additional weird thing about the current setup is that all of these cursors are meant to point _exactly_ at a specific entry in the encoded part, so even `peek` has a slightly weird semantics. Maybe the thing to do is add a new `Cursor::get(&self, part: EncodedPart) -> ...` function to `Cursor`? Taking `&self` by reference would make it more obvious that the cursor is not getting modified, and we can get rid of all these harmless-but-noisy clone calls...

Reply:
Pushed that idea up - does that feel any better?

Reply:
Looks great! Thanks for iterating :)